README.md
## Example Usage of CLI

```
lein uberjar

java -jar ./target/gr-parser-0.1.0-SNAPSHOT-standalone.jar pipe_delimitted pipe :date-of-birth

java -jar ./target/gr-parser-0.1.0-SNAPSHOT-standalone.jar pipe_delimitted pipe :last-name

java -jar ./target/gr-parser-0.1.0-SNAPSHOT-standalone.jar "pipe_delimitted" pipe "[:gender :last-name]"
```

## Running the API:

Starting the API server on http://localhost:8080

```
lein run -m gr-parser.web/-main
```

## Example Requests

### Adding Records
```
curl -i -H 'Content-Type: application/json' -XPOST 'http://localhost:8080/records' -d '{
  "delimitter": "pipe",
  "data": "LAST | FIRST | Female | BLUE | 1/1/1111"
}
'

curl -i -H 'Content-Type: application/json' -XPOST 'http://localhost:8080/records' -d '{
  "delimitter": "pipe",
  "data": "Bui | Chris | Male | BLUE | 5/18/1111"
}
'
```

### Sorted Records
```
curl -i  -XGET 'http://localhost:8080/records/gender'
curl -i  -XGET 'http://localhost:8080/records/birthdate'
curl -i  -XGET 'http://localhost:8080/records/name'
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-04-04
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-04-04
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/gr-parser/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/gr-parser/compare/0.1.0...0.1.1


test/gr_parser/web_test.clj
(ns gr-parser.web-test
  (:require [gr-parser.web :as sut]
            [clojure.test :refer :all]
            [cheshire.core :as json]
            [ring.mock.request :as mock]))

(deftest create-record-test
  (testing "People can be POST'd to the database with different delimitters"
    (let [people-db (atom [])
          handler (sut/make-app people-db)
          response (handler
                    (mock/request :post "/records" {:delimitter "pipe"
                                                    :data "B | C | Male | Blue | 5/18/1993"}))]
      (is (= 200 (:status response)))
      (is (= 1 (count @people-db)))
      (let [response (handler
                      (mock/request :post "/records" {:delimitter "comma"
                                                      :data "H, M, Female, Red, 3/3/1987"}))]
        (is (= 200 (:status response))))
      (is (= 2 (count @people-db)))
      (let [response (handler
                      (mock/request :post "/records" {:delimitter "space"
                                                      :data "C E Male Blue 2/10/1987"}))]
        (is (= 200 (:status response))))
      (is (= 3 (count @people-db)))
      (is (= [#gr_parser.core.Person{:last-name      "B",
                                     :first-name     "C",
                                     :gender         "Male",
                                     :favorite-color "Blue",
                                     :date-of-birth  #inst "1993-05-18T05:00:00.000-00:00"}
              #gr_parser.core.Person{:last-name      "H",
                                     :first-name     "M",
                                     :gender         "Female",
                                     :favorite-color "Red",
                                     :date-of-birth  #inst "1987-03-03T06:00:00.000-00:00"}
              #gr_parser.core.Person{:last-name      "C",
                                     :first-name     "E",
                                     :gender         "Male",
                                     :favorite-color "Blue",
                                     :date-of-birth  #inst "1987-02-10T06:00:00.000-00:00"}]
             @people-db)))))

(def db-state (atom [#gr_parser.core.Person{:last-name      "B",
                                            :first-name     "C",
                                            :gender         "Male",
                                            :favorite-color "Blue",
                                            :date-of-birth  #inst "1993-05-18T05:00:00.000-00:00"}
                     #gr_parser.core.Person{:last-name      "H",
                                            :first-name     "M",
                                            :gender         "Female",
                                            :favorite-color "Red",
                                            :date-of-birth  #inst "1987-03-03T06:00:00.000-00:00"}
                     #gr_parser.core.Person{:last-name      "C",
                                            :first-name     "E",
                                            :gender         "Male",
                                            :favorite-color "Blue",
                                            :date-of-birth  #inst "1987-02-10T06:00:00.000-00:00"}]))

(deftest get-records-gender-test
  (testing "Records are sorted by gender, female first, then by last-name"
    (let [handler   (sut/make-app db-state)
          resp-body (-> (handler (mock/request :get "/records/gender"))
                        (get :body)
                        (json/parse-string true))]
      (is (= [{:last-name      "H"
               :first-name     "M"
               :gender         "Female"
               :favorite-color "Red"
               :date-of-birth  "03/03/1987"}
              {:last-name      "B"
               :first-name     "C"
               :gender         "Male"
               :favorite-color "Blue"
               :date-of-birth  "05/18/1993"}
              {:last-name      "C"
               :first-name     "E"
               :gender         "Male"
               :favorite-color "Blue"
               :date-of-birth  "02/10/1987"}]
             resp-body)))))

(deftest get-records-name-test
  (testing "Records are sorted by last-name descending"
    (let [handler   (sut/make-app db-state)
          resp-body (-> (handler (mock/request :get "/records/name"))
                        (get :body)
                        (json/parse-string true))]
      (is (= [{:last-name      "H"
               :first-name     "M"
               :gender         "Female"
               :favorite-color "Red"
               :date-of-birth  "03/03/1987"}
              {:last-name      "C"
               :first-name     "E"
               :gender         "Male"
               :favorite-color "Blue"
               :date-of-birth  "02/10/1987"}
              {:last-name      "B"
               :first-name     "C"
               :gender         "Male"
               :favorite-color "Blue"
               :date-of-birth  "05/18/1993"}]
             resp-body)))))

(deftest get-records-birthdate-test
  (testing "Records are sorted by birthdate ascending"
    (let [handler   (sut/make-app db-state)
          resp-body (-> (handler (mock/request :get "/records/birthdate"))
                        (get :body)
                        (json/parse-string true))]
      (is (= [{:last-name      "C"
               :first-name     "E"
               :gender         "Male"
               :favorite-color "Blue"
               :date-of-birth  "02/10/1987"}
              {:last-name      "H"
               :first-name     "M"
               :gender         "Female"
               :favorite-color "Red"
               :date-of-birth  "03/03/1987"}
              {:last-name      "B"
               :first-name     "C"
               :gender         "Male"
               :favorite-color "Blue"
               :date-of-birth  "05/18/1993"}]
             resp-body)))))


test/gr_parser/core_test.clj
(ns gr-parser.core-test
  (:require [clojure.test :refer :all]
            [gr-parser.core :as sut]))

(def expected-record
  (sut/map->Person {:first-name     "Chris"
                    :last-name      "Bui"
                    :date-of-birth  #inst "1993-05-18T05:00:00.000-00:00"
                    :favorite-color "Blue"
                    :gender         "Male"}))

(deftest parse-line-pipe-test
  (is (= expected-record
         (sut/parse-line (:pipe sut/delimitters)
                         "Bui | Chris | Male | Blue | 5/18/1993"))))

(deftest parse-line-comma-test
  (is (= expected-record
         (sut/parse-line (:comma sut/delimitters)
                         "Bui, Chris, Male, Blue, 5/18/1993"))))

(deftest parse-line-space-test
  (is (= expected-record
         (sut/parse-line (:space sut/delimitters)
                         "Bui Chris Male Blue 5/18/1993"))))

(def test-people [{:first-name "C",
                   :last-name "B",
                   :gender "Male",
                   :favorite-color "Blue",
                   :date-of-birth #inst "1993-05-18T05:00:00.000-00:00"}
                  {:first-name "R",
                   :last-name "S",
                   :gender "Female",
                   :favorite-color "Green",
                   :date-of-birth #inst "1993-09-10T05:00:00.000-00:00"}
                  {:first-name "B",
                   :last-name "R",
                   :gender "Male",
                   :favorite-color "Orange",
                   :date-of-birth #inst "1982-02-15T06:00:00.000-00:00"}
                  {:first-name "S",
                   :last-name "G",
                   :gender "Female",
                   :favorite-color "Red",
                   :date-of-birth #inst "1994-01-10T06:00:00.000-00:00"}
                  {:first-name "E",
                   :last-name "C",
                   :gender "Male",
                   :favorite-color "Blue",
                   :date-of-birth #inst "1987-03-13T06:00:00.000-00:00"}])

(deftest sort-gender-last-name-test
  (is (= [{:first-name "S",
           :last-name "G",
           :gender "Female",
           :favorite-color "Red",
           :date-of-birth #inst "1994-01-10T06:00:00.000-00:00"}
          {:first-name "R",
           :last-name "S",
           :gender "Female",
           :favorite-color "Green",
           :date-of-birth #inst "1993-09-10T05:00:00.000-00:00"}
          {:first-name "C",
           :last-name "B",
           :gender "Male",
           :favorite-color "Blue",
           :date-of-birth #inst "1993-05-18T05:00:00.000-00:00"}
          {:first-name "E",
           :last-name "C",
           :gender "Male",
           :favorite-color "Blue",
           :date-of-birth #inst "1987-03-13T06:00:00.000-00:00"}
          {:first-name "B",
           :last-name "R",
           :gender "Male",
           :favorite-color "Orange",
           :date-of-birth #inst "1982-02-15T06:00:00.000-00:00"}]
         (sut/sort [:gender :last-name] test-people))
      "Females come first and each gender is sorted by last-name"))

(deftest sort-gender-last-name-test
  (is (= [{:first-name "B",
           :last-name "R",
           :gender "Male",
           :favorite-color "Orange",
           :date-of-birth #inst "1982-02-15T06:00:00.000-00:00"}
          {:first-name "E",
           :last-name "C",
           :gender "Male",
           :favorite-color "Blue",
           :date-of-birth #inst "1987-03-13T06:00:00.000-00:00"}
          {:first-name "C",
           :last-name "B",
           :gender "Male",
           :favorite-color "Blue",
           :date-of-birth #inst "1993-05-18T05:00:00.000-00:00"}
          {:first-name "R",
           :last-name "S",
           :gender "Female",
           :favorite-color "Green",
           :date-of-birth #inst "1993-09-10T05:00:00.000-00:00"}
          {:first-name "S",
           :last-name "G",
           :gender "Female",
           :favorite-color "Red",
           :date-of-birth #inst "1994-01-10T06:00:00.000-00:00"}]
         (sut/sort :date-of-birth test-people))
      "Sorted by date of birth"))

(deftest sort-gender-last-name-test
  (is (= [{:first-name "R",
           :last-name "S",
           :gender "Female",
           :favorite-color "Green",
           :date-of-birth #inst "1993-09-10T05:00:00.000-00:00"}
          {:first-name "B",
           :last-name "R",
           :gender "Male",
           :favorite-color "Orange",
           :date-of-birth #inst "1982-02-15T06:00:00.000-00:00"}
          {:first-name "S",
           :last-name "G",
           :gender "Female",
           :favorite-color "Red",
           :date-of-birth #inst "1994-01-10T06:00:00.000-00:00"}
          {:first-name "E",
           :last-name "C",
           :gender "Male",
           :favorite-color "Blue",
           :date-of-birth #inst "1987-03-13T06:00:00.000-00:00"}
          {:first-name "C",
           :last-name "B",
           :gender "Male",
           :favorite-color "Blue",
           :date-of-birth #inst "1993-05-18T05:00:00.000-00:00"}]
         (sut/sort :last-name test-people))
      "Sorted by last-name descending"))

(deftest person-formatter-test
  (is (= "05/18/1993"
         ((:date-of-birth sut/person-formatter)
          #inst "1993-05-18T05:00:00.000-00:00"))
      "Date of birth formats as dd/MM/yyyy"))

(deftest format-record-test
  (is (= {:first-name "C",
          :last-name "B",
          :gender "Male",
          :favorite-color "Blue",
          :date-of-birth "05/18/1993"}
         (sut/format-record (first test-people)))))

(deftest console-print-test
  (let [output (with-out-str (sut/console-print test-people))]
    (is (= " last-name first-name   gender       favorite-color        date-of-birth\n         B          C     Male                 Blue           05/18/1993\n         S          R   Female                Green           09/10/1993\n         R          B     Male               Orange           02/15/1982\n         G          S   Female                  Red           01/10/1994\n         C          E     Male                 Blue           03/13/1987\n"
           output))))


src/gr_parser/core.clj
(ns gr-parser.core
  (:require [clojure.string :as str]
            [clojure.edn :refer [read-string]])
  (:refer-clojure :exclude [sort])
  (:gen-class))

(def ^:const delimitters
  "A map of the record's type of delimitter to a regex to split the
  record"
  {:pipe  #" \| "
   :comma #", "
   :space #" "})

(def fields [:last-name :first-name :gender :favorite-color :date-of-birth])
(defrecord Person [last-name first-name gender favorite-color date-of-birth])

(def date-of-birth-format (java.text.SimpleDateFormat. "MM/dd/yyyy"))

(defn parse-date-of-birth
  [dob]
  (.parse date-of-birth-format dob))

(defn person
  "Constructs a Person with date-of-birth parsed from a map"
  [p]
  (-> p
      (update :date-of-birth parse-date-of-birth)
      (map->Person)))

(defn parse-line
  [delimitter-regex line]
  (->> (str/split line delimitter-regex)
       (zipmap fields)
       (person)))

(defmulti sort (fn [k people] k))

(defmethod sort [:gender :last-name]
  [_ people]
  (sort-by (juxt :gender :last-name) people))

(defmethod sort :date-of-birth
  [_ people]
  (sort-by :date-of-birth people))

;; last name descending
(defmethod sort :last-name
  [_ people]
  (reverse (sort-by :last-name people)))

(def person-formatter
  "Provides a mapping of field to a function that formats the field"
  {:date-of-birth
   (fn [dob]
     (.format date-of-birth-format dob))})

(defn format-record
  [person]
  (reduce-kv (fn [m k v]
               (let [transform (get person-formatter k identity)]
                 (assoc m k (transform v))))
             {}
             person))

(def console-format "%10s %10s %8s %20s %20s")

(defn print-formatted
  [format-str fields]
  (println (apply format format-str fields)))

(defn console-print
  [people]
  (print-formatted console-format (map name fields))
  (doseq [p people]
    (let [formatted-person (format-record p)]
      (print-formatted console-format
                       (for [f fields]
                         (get formatted-person f nil))))))

(defn -main
  [file-path delimitter user-sort]
  {:pre [(#{"pipe" "comma" "space"} delimitter)
         (#{"[:gender :last-name]" ":last-name" ":date-of-birth"} user-sort)]}
  (let [delimitter (keyword (read-string delimitter))
        user-sort (read-string user-sort)
        file (slurp file-path)
        lines (str/split file #"\n")]
    (->> lines
         (map (partial parse-line (delimitter delimitters)))
         (sort user-sort)
         (console-print))))


src/gr_parser/web.clj
(ns gr-parser.web
  (:require [compojure.core :refer :all]
            [cheshire.core :as json]
            [gr-parser.core :as gr-parser]
            [ring.adapter.jetty :as jetty]
            [ring.middleware.json :refer [wrap-json-params]]
            [ring.middleware.keyword-params :refer [wrap-keyword-params]]
            [ring.middleware.defaults :refer :all]))

(defn success-response [body]
  {:body (json/generate-string body)
   :status 200
   :headers {"Content-Type" "application/json"}})

(defonce people-db (atom []))

(defn make-app [db]
  (-> (routes
       (POST "/records" {{:keys [delimitter data]:as params} :params :as req}
             (let [delimitter (keyword delimitter)
                   person (gr-parser/parse-line (gr-parser/delimitters delimitter)
                                                data)]
               (swap! db conj person)
               (success-response "Ok")))
       (GET "/records/gender" []
            (success-response
             (map gr-parser/format-record
                  (gr-parser/sort [:gender :last-name] @db))))
       (GET "/records/birthdate" []
            (success-response
             (map gr-parser/format-record
                  (gr-parser/sort :date-of-birth @db))))
       (GET "/records/name" []
            (success-response
             (map gr-parser/format-record
                  (gr-parser/sort :last-name @db)))))
      (wrap-keyword-params)
      (wrap-json-params)
      (wrap-defaults api-defaults)))

(def wrapped-app (make-app people-db))

(defn -main []
  (jetty/run-jetty #'wrapped-app {:join? false :port 8080}))
